tools/wptrunner/wptrunner/testdriver-extra.js

```diff
@@ -76,6 +76,15 @@
             pending_resolve = resolve;
             pending_reject = reject;
         });
+        for (let actionSequence of actions) {
+            if (actionSequence.type == "pointer") {
+                for (let action of actionSequence.actions) {
+                    if (action.type == "pointerMove" && action.origin instanceof Element) {
+                        action.origin = {selector: get_selector(action.origin)};
```

Comment:
How do we handle when an origin element is out of the screen? I mean for click case we were scrolling into view to container the target. How about here?

Reply:
Actions don't do that. There was a plan to add explicit scrolling to the actions spec, so you could (in terms of this api) do something like

var elem = document.getElementById("element");
Actions()
  .scroll(0, 0, origin=elem)
  .pointerDown(elem)
  .pointerUp()
  .send()
```

But at the moment that doesn't work and you need to call `scrollIntoView` before you send actions.

Reply:
One thing that makes our job a little harder and possibly inconsistent across different implementation is calculating "in-view center point of element" which I believe now we should do in testdriver-vendor.js as well (at least in Chrome side) . I was hoping that this calculation can be done in this layer (which we already have functions for and used them in click) and only x's/y's are passed over.

Reply:
*process* doesn't do anything, as such, it builds a data stucture that's passed to *dispatch*.

Not supporting `origin` is probably a bad idea since the 80% use case here is clicking on a specific element. We could move the support into the client, but notice that there isn't a roundtrip and the semantics are equivalent. For example if I arrange a set of actions that click on element a, which causes scrolling, then click on element b, the coordinates of the element change between the command being sent and the actual action executing. But because we resolve the coordinates right before the click, things  work with the WebDriver method but don't with the precompute method. Obviously there are ways to break both cases e.g. animating the elment.

Reply:
I agree with the ease of use in all the scenarios.
I believe if the page wants to move around the layout during the actions their best bet is to break off the actions into two calls as they need to make sure the action is indeed done and they verity the effect (e.g. scrolling) before injecting the next action. Otherwise they are just writing a racy test and they are not sure what action is done in what state of the page.

Aside from that what is the driver supposed to do if it gets an element that is outside of the view? There is no scrollIntoView at this point in any steps at this point and not a clear error message. I'm just trying to make sure not to leave these steps up to each implementation as it is just another avenue for interop problems.

Reply:
Just to be clear I'm not saying webdriver spec should change or anything. I just think that if we expose some minimal set in test_driver that is less prune to differences in the implementation of vendors in their drivers it makes it so that wpt is easier to develop. Any builder or wrapper api like calculating the coordinate from a target and what not that to makes the life of the test developer easier could all live here in wpt repo and shared across browsers so we don't run into interop issues on that part.
This can also apply to that one action in on tick. It is still possible to do anything across multiple ticks since ticks have duration.

Reply:
Well the steps aren't up to each implementation; the reason that this is following WebDriver is that we're fuly reusing the WebDriver-specified semantics. I agree that there's a chance of errors in edge cases when people [re]implement the WebDriver logic.

It seems like the request here is effectively to disallow `origin=pointer` in this API and force everyone to call a js method for getting the element coordinates; I would be very *very* reluctant to allow passing an element in with the actions sequence but that element having its coordinates evaluated at a different time to that specified in the WD spec; it would be extremely confusing. So then the question is whether it's more reasonable to make authors write `.pointerMove(...testdriver.elementCoordinates(element), origin="viewport")` or allow them to write `.pointerMove(0, 0, origin=element)` and take increased risk of some implemenation having an error.

Re ticks, I just realised why that can't work (i.e. why webdriver has the design it has). If you want to simulate a pinch action, you require two fingers to move simultaneously. That means multiple `pointerMove` events in a single tick (the implementation will act to make the two movements interleave in the way you would expect).

Reply:
Re: origin, I made the same suggestion; I was just pointing out that there is a serious burden on test authors to having to do all the position compoutations vs letting the implementation do it. Typically we don't prioritise implementation simplicity over authoring experience, so I"m reluctant to do so in this case, but if you are claiming that Chrome can't implement the API if we allow elements to be passed in, I guess we have to put the burden on authors. The alternative is we come up with a testsuite that ensures the 

The way it works in WebDriver, the move duration is not orthogonal to the tick duration; the length of the tick is extended to be the maximum duration of any action in that tick. Otherwise you could have nonsense like a single pointer doing two different overlapping moves. Forcing all guestures to be build out of a series of moves corresponding to different ticks seems like it doesn't work because you have to spin the event loop enough at the end of a tick to ensure that events generated in that tick have a chance to be processed, so you can't guarantee timing.

Generally I feel like this conversation is going through the same process that led to the current WebDriver actions API. Whilst there is some complexity in that API it is motivated by real requirements and I'm not sure we're going to find out that there's a simpler yet compatible model that can do all the same things.

Reply:
Sure. Sounds good then. Let's go with your suggestion. @foolip  is it possible for us to run these tests added here on try bots to make sure we did things correctly in testdriver-vendor.js? These are a little different from the py tests for webdriver and are very similar to our layout tests.

Reply:
You would have to copy over the changes manually and see how they work, we don't have a way of created a Chromium CL from a WPT PR, the only thing we do in that direction is import master continuously, *after* the PR is merged.